### ISSUE TITLE
MBT with StateMachine: Information from the SUT to the model or unique IDs for operation instances

### ISSUE THREAD

**Issue by helinko** [fscheck/FsCheck#366]
I am modeling CRUD-like operations on a project. So the operations create, read, update and delete objects in a database. 

I have the problem of mapping an object from the SUT to an object in my model. For example, I'd like to search for an arbitrary object in the SUT and check that it has the same properties as it's corresponding object in the model.

I can't use the ID that SUT gives me, because at the point of creating the actual object the model is inaccessible. Another idea was to add another (predetermined) unique ID to the objects, but I couldn't come up with an elegant way to implement it. I guess I would need to give the ID as a parameter to the Create operation, but each instance of that would be allowed to run only once. I would also have to give a limit to the amount of different possible Create operations.

I have to admit I haven't quite grasped the generators yet, so there might be a simple solution relating to them. Any ideas?

**Comment by kurtschelfthout** [comment]
Are you using the `Experimental.StateMachine` API?

**Comment by helinko** [comment]
Yes.

**Comment by kurtschelfthout** [comment]
Have a look at the `OperationResult` class. Basically it allows you in the model to assign a "variable" - a placeholder for the eventual result of an operation. And then in the `Check` method you can assign the real result coming from the SUT. So for example, in create's model you could just instantiate a new OperationResult as a placeholder for the id and keep it as part of the model, and then in the `Check` method have the SUT generate the actual id, and assign it to the `OperationResult`. Further on in your model you can then access the id via the `OperationResult` which is passed down as part of the model.

Not sure if my rambling explanation makes sense, there is an example here: https://github.com/fscheck/FsCheck/blob/master/src/FsCheck/Script.fsx#L47

Where it is used to represent the process id generated by spawning a new process.

Unfortunately this is not documented anywhere, and probably a bit rough around the edges.

**Comment by helinko** [comment]
I'm now able to set the ID in the model, but I'm not sure how I can generate the fresh instances. I tried returning this in Next: 

`Gen.Fresh<Operation<SUTState, ModelState>>( () => new CreateMinimalObject( justSomeObjType ) )`

But it seems to lead to a sequence which repeats same instance of CreateMinimalObject. 

I'd also like it to generate a few different kinds of operation instances. I previously had a list like `{new CreateMinimalObject(type1), new CreateMinimalObject(type2), ... }`, and then I would make it into a generator with elements. So how could I have it generate those few types and always fresh instances?

**Comment by kurtschelfthout** [comment]
I would expect you to be able to just use `Gen.elements [ new CreateMinimalObject(t1); new CreateMinimalObject(t2);...]` as a result from `Next` but those are just the operations. The actual objects would be created in the `CreateMinimalObject` class in the `Check` method, so maybe something is going wrong there.

**Comment by helinko** [comment]
I managed to get it working with OperationResult. Thanks!

It would be nice to have a prettier way to get information from the SUT to the model. I also had a similar problem in Setup, and ended up having a private field in the Spec class which I then initialized from the SUT in Setup.Model(). So basically I set up and tear down my SUT in Model(), and then set it up again in Actual(). I also have to do my own book-keeping for it. 

I guess most solutions for that are not viable because a full trace is created for the model first, and therefore you don't know anything about the SUT while running the model. So why was that sequence of calls chosen?

This might not be a direct (or good) solution, but referring to those fresh instances I was talking about earlier: I'd like the operation instances to be separate, even the ones that have the same parameters. So I could, for example in Run, set a property/field for the Operation that I can then read later in Check. If I do this currently, it will re-set the property in the next call for Run that has the same parameters for the parent Operation.

I could of course put that property in as a parameter, but as far as I understand, I can't have an infinite set that the generator picks from.

It might be wasteful to do that always, but maybe it could be done with some specific generator?

**Comment by kurtschelfthout** [comment]
> It would be nice to have a prettier way to get information from the SUT to the model.

I think it's going to be hard to talk about without showing some specific example...what you describe certainly sounds cumbersome but I'm not convinced it's the best or only way to do things.

> So why was that sequence of calls chosen?

This ensures that the model encodes enough information to be able to do shrinking without having to run the SUT itself. I.e. the assumption is that the SUT is side-effecting and expensive to run, so you only want to run sequences that are "good" according to the model. And you want to be able to do some "exploring" with it, e.g. start with an existing sequence and remove or add an operation in the middle, without having to run the full SUT up until that point.

It's often useful in the model to refer to the result of an earlier operation in the sequence "symbolically" without having to store the actual value. In addition, this gives the model more information because it encodes dependencies between operations. For example, if the shrinker tries to remove an operation that contains an assignment to an OperationResult, it also removes any later operations that read from that OperationResult.

> It might be wasteful to do that always, but maybe it could be done with some specific generator?
I don't understand your question. A code example would help a lot here I think.

**Comment by helinko** [comment]
To avoid confusion, this message is only about Operations and "data-flow" between the Run and Check (or model and SUT). 

Example: The model’s state is a sequence of elements, all of which are unique. I want to implement a Delete operation that deletes a random element. How can I choose an element randomly, while also getting the information in Check which element was deleted? Maybe it’s something about the generators that I’m missing.

The “fresh instances” would solve this as follows: I would save the deleted element in the operation instance (as a property/field), which would let me refer to it in Check and delete the corresponding element in the SUT. So the state of operation instance would have to be separate from every other delete operation.

So I don’t understand the generators well enough to know if this is at all possible, but Next would then return a generator that would, in the simplest case, be formed with only ‘new Delete()’. But every time Delete.Model() is called during the test sequence, it’s a new instance of Delete. Am I making any sense here?

**Comment by kurtschelfthout** [comment]
> Example: The model’s state is a sequence of elements, all of which are unique. I want to implement a Delete operation that deletes a random element. How can I choose an element randomly, while also getting the information in Check which element was deleted? Maybe it’s something about the generators that I’m missing.

You would store a list of `OperationResult` values in your model, called e.g. `createdElements` of which you pick one to call delete on when you create a `Delete` operation.

I don't think you want to have information from the SUT creep into your model; once the model starts using values from the SUT directly (or the current test run) it becomes hard to cleanly separate the two, and there'd be many opportunities to fool yourself...

**Comment by helinko** [comment]
Turns out that returning Gen.Fresh in Next was what I needed all along. I just made a mistake when I tried it earlier, which lead me to rule it out prematurely. 

So I think I don't need to use OperationResult or any other way to get info from SUT to model, because now I can just choose randomly in Run and see the choice in Check.
